subprojects/core/src/main/java/org/gradle/internal/build/PlannedNodeGraph.java

```diff
@@ -0,0 +1,235 @@
+        private final Map<Node, NodeIdentity> nodeIdentityCache = new IdentityHashMap<>();
+
+        public Collector(ToPlannedNodeConverterRegistry converterRegistry) {
+            this.converterRegistry = converterRegistry;
+            this.detailLevel = DetailLevel.from(converterRegistry.getConvertedNodeTypes());
+        }
+
+        public DetailLevel getDetailLevel() {
+            return detailLevel;
+        }
+
+        public void collectNodes(Collection<Node> nodes) {
+            for (Node node : nodes) {
+                ToPlannedNodeConverter converter = converterRegistry.getConverter(node);
+                if (converter != null && converter.isInSamePlan(node)) {
+                    List<? extends NodeIdentity> nodeDependencies = findNodeDependencies(node);
+                    PlannedNodeInternal plannedNode = converter.convert(node, nodeDependencies, () -> findTaskDependencies(node));
+                    plannedNodes.add(plannedNode);
+                }
+            }
+        }
+
+        public PlannedNodeGraph getGraph() {
+            return new PlannedNodeGraph(detailLevel, plannedNodes);
+        }
+
+        private List<? extends NodeIdentity> findNodeDependencies(Node node) {
+            return findDependencies(node, Node::getDependencySuccessors, it -> true);
+        }
+
+        private List<TaskIdentity> findTaskDependencies(Node node) {
+            @SuppressWarnings("unchecked")
+            List<TaskIdentity> dependencies = (List<TaskIdentity>) findDependencies(node, Node::getDependencySuccessors, TaskIdentity.class::isInstance);
+            return dependencies;
+        }
+
+        private List<? extends NodeIdentity> findDependencies(
+            Node node,
+            Function<? super Node, ? extends Collection<Node>> traverser,
+            Predicate<NodeIdentity> identityFilter
+        ) {
+            return computeDependencies(traverser, n -> getNodeIdentityOrNull(n, identityFilter), node);
+        }
+
+        private NodeIdentity getNodeIdentityOrNull(Node node, Predicate<NodeIdentity> identityFilter) {
+            ToPlannedNodeConverter converter = converterRegistry.getConverter(node);
+            if (converter == null) {
+                return null;
+            }
+
+            // Cache all identities even if the current predicate will filter them out afterwards
+            NodeIdentity nodeIdentity = nodeIdentityCache.computeIfAbsent(node, converter::getNodeIdentity);
+            return identityFilter.test(nodeIdentity) ? nodeIdentity : null;
+        }
+    }
+
+    /**
+     * Computes dependencies of a node in breadth-first order, stopping at each dependency for which identity is provided.
+     */
+    private static <T> List<NodeIdentity> computeDependencies(
+        Function<? super T, ? extends Collection<? extends T>> traverser,
+        Function<T, NodeIdentity> maybeIdentityProvider,
```

Comment:
🤔  Do you think it makes sense to create two functional interfaces here instead of using `Function` to make the code better to understand?

Reply:
Done